server/sonar-process/src/main/java/org/sonar/process/systeminfo/SystemInfoHttpServer.java

```diff
@@ -0,0 +1,102 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.process.systeminfo;
+
+import fi.iki.elonen.NanoHTTPD;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.Properties;
+import org.slf4j.LoggerFactory;
+import org.sonar.process.DefaultProcessCommands;
+import org.sonar.process.systeminfo.protobuf.ProtobufSystemInfo;
+
+import static java.lang.Integer.parseInt;
+import static org.sonar.process.ProcessEntryPoint.PROPERTY_PROCESS_INDEX;
+import static org.sonar.process.ProcessEntryPoint.PROPERTY_SHARED_PATH;
+
+/**
+ * This HTTP server exports data required for display of System Info page (and the related web service).
+ * It listens on loopback address only, so it does not need to be secure (no HTTPS, no authentication).
+ */
+public class SystemInfoHttpServer {
+
+  private static final String PROTOBUF_MIME_TYPE = "application/x-protobuf";
+
+  private final Properties processProps;
+  private final List<SystemInfoSectionProvider> sectionProviders;
+  private final SystemInfoNanoHttpd nanoHttpd;
+
+  public SystemInfoHttpServer(Properties processProps, List<SystemInfoSectionProvider> sectionProviders) throws UnknownHostException {
+    this.processProps = processProps;
+    this.sectionProviders = sectionProviders;
+    InetAddress loopbackAddress = InetAddress.getByName(null);
+    this.nanoHttpd = new SystemInfoNanoHttpd(loopbackAddress.getHostAddress(), 0);
+  }
+
+  public void start() {
+    try {
+      nanoHttpd.start();
+      registerHttpUrl();
+    } catch (IOException e) {
+      throw new IllegalStateException("Can not start local HTTP server for System Info monitoring", e);
+    }
+  }
+
+  private void registerHttpUrl() {
+    int processNumber = parseInt(processProps.getProperty(PROPERTY_PROCESS_INDEX));
+    File shareDir = new File(processProps.getProperty(PROPERTY_SHARED_PATH));
+    try (DefaultProcessCommands commands = DefaultProcessCommands.secondary(shareDir, processNumber)) {
+      String url = getUrl();
+      commands.setSystemInfoUrl(url);
+      LoggerFactory.getLogger(getClass()).debug("System Info HTTP server listening at {}", url);
+    }
+  }
+
+  public void stop() {
+    nanoHttpd.stop();
+  }
+
+  // visible for testing
+  String getUrl() {
+    return new StringBuilder().append("http://").append(nanoHttpd.getHostname()).append(":").append(nanoHttpd.getListeningPort()).toString();
```

Comment:
you could write this with `+` operators, the compiler will create the `StringBuilder` for you and readability might be better